maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java

```diff
@@ -1624,20 +1635,65 @@ static void populateProperties(
         String mavenBuildVersion = CLIReportingUtils.createMavenVersionString(buildProperties);
         systemProperties.setProperty("maven.build.version", mavenBuildVersion);
 
-        BasicInterpolator interpolator =
-                createInterpolator(paths, systemProperties, userProperties, userSpecifiedProperties);
-        for (Map.Entry<Object, Object> e : userSpecifiedProperties.entrySet()) {
-            String name = (String) e.getKey();
-            String value = interpolator.interpolate((String) e.getValue());
-            userProperties.setProperty(name, value);
-            // ----------------------------------------------------------------------
-            // I'm leaving the setting of system properties here as not to break
-            // the SystemPropertyProfileActivator. This won't harm embedding. jvz.
-            // ----------------------------------------------------------------------
-            if (System.getProperty(name) == null) {
-                System.setProperty(name, value);
-            }
+        // ----------------------------------------------------------------------
+        // Options that are set on the command line become system properties
+        // and therefore are set in the session properties. System properties
+        // are most dominant.
+        // ----------------------------------------------------------------------
+
+        Properties userSpecifiedProperties =
+                commandLine.getOptionProperties(String.valueOf(CLIManager.SET_USER_PROPERTY));
+        userProperties.putAll(userSpecifiedProperties);
+
+        // ----------------------------------------------------------------------
+        // Load config files
+        // ----------------------------------------------------------------------
+        Function<String, String> callback =
+                or(paths::getProperty, prefix("cli.", commandLine::getOptionValue), systemProperties::getProperty);
```

Comment:
Please also add: `userProperties::getProperty`, so we can use: `mvn ... -Dname=xxx`

Reply:
Good catch!

Reply:
This should not be necessary.   The command line options are added to the `userProperties` a few line earlier:
https://github.com/apache/maven/blob/c4fa7a099a7f338e5c78b93fb476f9d451349dda/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java#L1644-L1646
So all those user-provided properties will be interpolated the same way than those from the properties file (and thus be available for being themselves values for interpolation).

Reply:
it is necessary

Reply:
Properties for testing: 
```
${includes}=default.properties,env-${envName}.properties
```

command:
```
mvn -v -DenvName=test
```

fail with:

```
java.nio.file.NoSuchFileException: ..../repos/test/test/.mvn/env-${envName}.properties
```